### ISSUE TITLE
Interception of created navigables

### ISSUE THREAD

**Issue by OrKoN** [w3c/webdriver-bidi#756]
Many WebDriver BiDi methods allow configuring the browser behavior globally or per existing navigable. For example, `network.addIntercept` can be used either per navigable(s) or apply globally. Similarly, event subscriptions can be done globally or per navigable. The important requirement for calling these methods for a navigable is that the navigable has to exist.

There is another use case that is currently not covered: how to configure newly created navigables? For example, if you want to subscribe to a subset of events depending on the navigable parameter, there is no way to do it in a race-free way without reloading the newly created navigable. Another example could be network interception and event logging: there could be many navigables in a browser that you do not want to intercept requests in, and dynamically created navigables that you do want to set up interception in. 

We could extend the navigable creation algorithm with an ability to pause loading of a navigable, notify the client about the navigable and allow the client to resume loading once the commands configuring the behavior of the newly created navigable have been sent. Ideally, that should extend to worker realms since many workers might not be interesting for interception.

cc @sadym-chromium

**Comment by OrKoN** [comment]
At least a couple of feedback items from Playwright in https://github.com/w3c/webdriver-bidi/issues/769 could be solved by this feature. cc @sadym-chromium

**Comment by whimboo** [comment]
I've added that issue to the TPAC agenda for discussion.

**Comment by dmurph** [comment]
this would be incredibly useful for testing navigation capturing for installed web applications.

**Comment by whimboo** [comment]
> <simonstewart> Action: Julian to see if gecko can pause creation of new navigables

I wonder if [`document.blockParsing()`](https://searchfox.org/mozilla-central/rev/5a4f7aa7b8cafa066ef06ec9bf09238375e74b02/dom/webidl/Document.webidl#472) could be helpful here. But I'm not sure if that is just for parsing scripts, or if we could stop the full parser.

@juliandescottes do you know?

**Comment by whimboo** [comment]
We had some internal discussions today, and one option that emerged is allowing customization of specific user contexts, as mentioned in [this issue](https://github.com/w3c/webdriver-bidi/issues/789). This approach would still require opening an additional tab but seems much cleaner than returning early from `browsingContext.createContext` and then resuming.

@OrKoN would that be an option, or won't that cover all specific features wanted for that request?

**Comment by OrKoN** [comment]
I think it would not work for some existing Puppeteer's APIs, e.g., the request interception in Puppeteer is per top-level browsing context and would not be easy to apply to the user context. 

I think instead of interception (given the complexity and the requirement for message roundtrips) we could introduce a parameter to indicate that a particular configuration (emulation or event subscription) needs to be inherited by the contexts opened from the current one. I think this Playwright feedback item would be addressed by either interception or inheritance of certain configuration options:

>  Provide a way to configure the viewport before a popup or new tab starts loading content. Currently, the browsingContext.SetViewport command can only be sent to an existing page, and there is no way to pause the load of a new page (e.g., opened via ctrl+click or a popup); these pages always load with the default viewport first. 

(https://github.com/microsoft/playwright/issues/32577)

**Comment by OrKoN** [comment]
So if we do not want to go with pausing the loading, we could allow inheriting configurations on the per-command basis, for example, https://github.com/w3c/webdriver-bidi/pull/803

**Comment by whimboo** [comment]
@OrKoN a couple of APIs currently already support the configuration before the created navigable is loaded. Do you think that we should keep this issue open or can we close it as done?

**Comment by OrKoN** [comment]
Yeah, I think it can be closed for now as current client requests are satisfied.
